### DISCUSSION TITLE
monitor.dailystats anomalies

### DISCUSSION THREAD

**Discussion by gyveri** [ZuinigeRijder/hyundai_kia_connect_monitor#62] [Q&A]
Thank you for your advice regarding not exceeding the api access limitation. I have devised a daily schedule that provides the data I  need without exceeding this limitation. Since I started using the schedule, run_monitor_once.sh appears to run consistently and updates the monitor.csv file and the monitor spreadsheet with data that makes sense.

That said, I’m curious as to what is meant in README.md by “only available for Europe” with reference to monitor.dailystats.csv and monitor.tripinfo.csv. run_monitor_once.sh appears to update monitor.dailystats.csv and the monitor.dailystats spreadsheet, but appears not to update monitor.tripinfo.csv. Moreover, although some of the data with which monitor.dailystats.csv is updated makes sense, the data for each trip in a given day is repeated at hourly intervals from the time at which it first appears until data for the first trip of the next day appears (and repeats). Also the monitor.dailystats spreadsheet only displays data for the last trip of the day, no matter how short.  So it looks like the dailystats data is available in the U.S, but monitor.dailystats.py is not handling it correctly. 
Can this be fixed?

**Comment by ZuinigeRijder** [comment]
I will have a look in the source code HyundaiBlueLinkAPIUSA.py to see what is available in hyundai_kia_connect_api. In the mean time, can you post a screenshot of monitor.dailystats spreadsheet and also the standard output of running dailystats.py? So I better understand what is provided in USA in combination with the source code.

**Reply by gyveri** [reply]
I've attached a .pdf file of the most recent part of the monitor.dailystats spreadsheet and a copy of the monitor.dailystats.csv file. In the real world the car did the following trips
8-16 - did not drive the car, but turned it on twice to fix premature service reminder
8-15 - 3 round trips, 6.1 miles, 3.7 miles. 0.6 miles (all distances are each-way)
8-14 - 1 round trip, 41.5 miles
8-13 - 1 round trip, 82 miles
8-12 - 1 round trip , 3.7 miles
Please let me know if you have questions or need copies of additional files
[monitor.dailystats.csv](https://github.com/user-attachments/files/16646155/monitor.dailystats.csv)
[monitor.dailystats - monitor.dailystats-1.pdf](https://github.com/user-attachments/files/16646158/monitor.dailystats.-.monitor.dailystats-1.pdf)

**Comment by ZuinigeRijder** [comment]
I looked at the source code and also at the PDF and monitor.dailystats.csv.
The following problems are in hyundai_kia_connnect_api (so not in hyundai_kia_connect_monitor).
1. vehicle.daily_stats is not sorted, only sorted in KiaUvoApiEU.py (line 1249 till 1251). therefore double (repeated) items each day
2. method update_day_trip_info is NOT overruled/implemented in HyundaiBlueLinkAPIUSA.py

For 1. this could be fixed by sorting when assigning this to vehicle.daily_stats (via setter) instead of doing this at each implementation

For 2. this could be fixed by figuring out the information in the bluelink API, e.g. by execution hyundai_kia_connect_monitor.debug.py and see if the trip information is available in the returned data for HyundaiBlueLinkAPIUSA.py.  Note that you could look at the implementation of method update_day_trip_info in KiaUvoApiEU.py, KiaUvoApiAU.py and KiaUvoApiCN.py. But then the API has to be figured out, like in method _get_trip_info in the previous mentioned implementations. 
``
url = self.SPA_API_URL + "vehicles/" + vehicle.id + "/tripinfo"
``

> Also the monitor.dailystats spreadsheet only displays data for the last trip of the day, no matter how short.

This is because the trip data is other information, the dailystats only shows the last entry of a day. Trip info must be implemented in update_day_trip_info  to have the trip information.

Hope this makes sense for you.

**Comment by gyveri** [comment]
Many thanks for your observations regarding the cause of the anomalies. 
As I have noted before, the Python code in this app is way more sophisticated than anything I have written. I would therefore hesitate to modify the code to implement your suggestion. So ... would be willing to do this? I will be happy to test anything your write.

**Comment by ZuinigeRijder** [comment]
For problem 1 I made a fix, based on hyundai_kia_connect_api-3.22.7 by sorting in the setter in Vehicle.py. See attached the zip with the changes.

[hyundai_kia_connect_api.zip](https://github.com/user-attachments/files/16688948/hyundai_kia_connect_api.zip)

Just unpack and move the files in hyundai_kia_connect_api  to hyundai_kia_connect_monitor.hyundai_kia_connect_api 
Let me know if this solves the problem, that the sequence of trips on one day is repeated in monitor.dailystats.csv onwards (after the update).

For problem 2 we need further discussion how we can find the api for the trip information (if available).

**Reply by gyveri** [reply]
New hyundai_kia_connect_api installed. Leaving on 2-leg trip soon.

**Comment by gyveri** [comment]
Well, the trip turned out to have 4 legs:
1. Started 09:25, drove 8 miles, arrived before 10
2. 10:56 - momentary key on. 0 distance
3. Started 10:57, drove 0.6 miles arrived after 11
4. Started  11:31, drove 8 miles, arrived before noon
monitor.dailystats.csv saved at 14:00 is attached
[monitor.dailystats.csv](https://github.com/user-attachments/files/16699931/monitor.dailystats.csv)

**Comment by ZuinigeRijder** [comment]
Seems that my changes did not work, there is still a repetition in the data.....

````
20240821 10:00, 9, km, 1455, 1521,  1234, 101, 120, 0
20240821 11:00, 0, km, 28, 0,  0, 8, 20, 0
20240821 12:00, 9, km, 1455, 1521,  1234, 101, 120, 0
20240821 12:00, 0, km, 28, 0,  0, 8, 20, 0
20240821 12:00, 0, km, 175, 168,  108, 27, 40, 0
20240821 12:00, 8, km, 1927, 897,  1689, 128, 110, 0
20240821 13:00, 9, km, 1455, 1521,  1234, 101, 120, 0
20240821 13:00, 0, km, 28, 0,  0, 8, 20, 0
20240821 13:00, 0, km, 175, 168,  108, 27, 40, 0
20240821 13:00, 8, km, 1927, 897,  1689, 128, 110, 0
20240821 14:00, 9, km, 1455, 1521,  1234, 101, 120, 0
20240821 14:00, 0, km, 28, 0,  0, 8, 20, 0
20240821 14:00, 0, km, 175, 168,  108, 27, 40, 0
20240821 14:00, 8, km, 1927, 897,  1689, 128, 110, 0
````

Is the sorting not working? Can you mail me the output of "python debug.py" to my gmail account zuinige.rijder, because debug logging shows if it sorts and the sorting before and after.

**Comment by ZuinigeRijder** [comment]
Another IONIQ 5 USA user did mail me a debug.log file and I see that the dailystats API returns different information than in Europe.
In Europe only the latest server values of a day are returned, In USA multiple server values of a day are returned, including the timestamp.

More specific, in USA you have daily stats information for all trips (so no cumulative value, but for each trip the actual values), in Europe only the cumulative latest value per day. 

I was expecting only one dailystats value per day in monitor.py. So besides the sorting, something else has to be done. I need to think for an elegant solution......because the data is really different if you talk about cumulative data on a day (Europe) or for each trip the exact data (USA).

One simple solution is to also compute the cumulative values for USA in HyundaiBlueLinkAPIUSA.py, so the returned data is the same. Actually the DailyStats USA data is richer than in Europe, but in Europe we have separate trip data call with start and end-time and other information. In DailyStats USA it is a combination, there is a start date and duration is in the returned data, including speed, etc.. So probably with the information returned in USA you could compose DailyStats and trip information........but this is complex.

Maybe we should move this discussion to hyundai_kia_connect_api, so others can also contribute?

**Comment by gyveri** [comment]
Yes, let's move this to gitHub (without the debug log files I and the other Ioniq 5 user uploaded).
I suppose that this means that trip information is available in the U.S. after all - it's just in a different place than in Europe.
I wonder why Hyundai goes to all the trouble of writing significantly different apis for different markets.
--Ian

    On Thursday, August 22, 2024 at 01:01:22 PM PDT, Zuinige Rijder ***@***.***> wrote:   

 


Another IONIQ 5 USA user did mail me a debug.log file and I see that the dailystats API returns different information than in Europe.
In Europe only the latest server values of a day are returned, In USA multiple server values of a day are returned, including the timestamp.

More specific, in USA you have daily stats information for all trips (so no cumulative value, but for each trip the actual values), in Europe only the cumulative latest value per day.

I was expecting only one dailystats value per day in monitor.py. So besides the sorting, something else has to be done. I need to think for an elegant solution......because the data is really different if you talk about cumulative data on a day (Europe) or for each trip the exact data (USA).

One simple solution is to also compute the cumulative values for USA in HyundaiBlueLinkAPIUSA.py, so the returned data is the same. Actually the DailyStats USA data is richer than in Europe, but in Europe we have separate trip data call with start and end-time and other information. In DailyStats USA it is a combination, there is a start date and duration is in the returned data, including speed, etc.. So probably with the information returned in USA you could compose DailyStats and trip information........but this is complex.

Maybe we should move this discussion to hyundai_kia_connect_api, so others can also contribute?

**Comment by ZuinigeRijder** [comment]
Discussion continues with [this issue in hyundai_kia_connect_api.](https://github.com/Hyundai-Kia-Connect/hyundai_kia_connect_api/issues/617)
